# Rossmann PostgreSQL Analysis
**Author:** Priyanshu Singh | Data Analyst
**Tools:** PostgreSQL, Python, SQLAlchemy

## Objective
Analyze 844,338 Rossmann retail sales records using advanced SQL.
Demonstrates CTEs, window functions, JOINs and aggregations.

## Queries Covered
1. Top 10 stores by average sales
2. Promotion impact on sales
3. Revenue by store type (JOIN)
4. Monthly sales trend
5. Store rankings using RANK()
6. Month-over-month growth using LAG()
7. Above average stores using CTE
8. Year-over-year growth — raw totals (reference)
9. Year-over-year growth — corrected for partial year (avg daily sales)
10. Store performance tier classification (CASE WHEN)
11. Tier distribution summary (window function on aggregate)
12. Q4 promotional effectiveness by store type (conditional aggregation)

In [1]:
from dotenv import load_dotenv
import os
from sqlalchemy import create_engine
import pandas as pd

load_dotenv()
password = os.getenv('DB_PASSWORD')

engine = create_engine(f'postgresql://postgres:{password}@localhost:5432/rossmann')
print("Connected successfully")

Connected successfully


## Top 10 Stores by Average Daily Sales
Identifies the highest-performing stores across the 1,115-store chain
based on average daily sales — a fairer metric than total revenue,
as it accounts for stores with different numbers of active trading days.

In [2]:
q1 = pd.read_sql("""
SELECT
    store,
    ROUND(AVG(sales)::numeric, 2) AS average_sales,
    SUM(sales) AS total_revenue,
    COUNT(DISTINCT date) AS active_days
FROM rossmann_sales
GROUP BY store
ORDER BY average_sales DESC
LIMIT 10""", engine)

print("Top 10 Stores by Average Daily Sales")
print(q1)

Top 10 Stores by Average Daily Sales
   store  average_sales  total_revenue  active_days
0    817       21757.48     17057867.0          784
1    262       20718.52     19516842.0          942
2   1114       20666.56     16202585.0          784
3    251       19123.07     14896870.0          779
4    842       18574.80     11553523.0          622
5    513       18179.09     14252406.0          784
6    562       17969.56     16927322.0          942
7    788       17961.91     14082141.0          784
8    383       17294.72     13489879.0          780
9    756       16574.82     12911782.0          779


**Finding:** Store 817 leads the chain with €21,757 average daily sales —
3.1x the chain-wide average of €6,955. All top 10 stores exceed €16,500,
suggesting a small cluster of flagship stores drives disproportionate revenue.

## Promotion Impact
Compares average sales on promotion vs non-promotion days.

In [3]:
q2 = pd.read_sql("""
select promo,
    round(avg(sales)::numeric,2) as average_sales,
    count (*) as total_days
from rossmann_sales
group by promo
order by average_sales desc""",engine)

print("Impact of Promo on the Sales")
print(q2)

Impact of Promo on the Sales
   promo  average_sales  total_days
0      1        8228.74      376875
1      0        5929.83      467463


### The impact of promotion affectively affects the sales by more than 30 percent

## Revenue by Store Type
Uses JOIN to combine sales and store metadata. Compares avg and total revenue across store types.

In [4]:
q3 = pd.read_sql("""
select 
    s.storetype,
    round(avg(r.sales)::numeric,2) as average_sales,
    sum(r.sales) as total_sales
from rossmann_sales r 
    join rossmann_store s on r.store = s.store
group by s.storetype
order by average_sales desc""",engine)

print("Average Sales per Store Type")
print(q3)

Average Sales per Store Type
  storetype  average_sales   total_sales
0         b       10233.38  1.592314e+08
1         c        6933.13  7.832214e+08
2         a        6925.70  3.165335e+09
3         d        6822.30  1.765393e+09


Storetype "b" has higher average sales than any other store type.

## Monthly Sales Trend
Identifies seasonal patterns across 12 months.

In [5]:
q4 = pd.read_sql("""
select 
    month,
    round(avg(sales)::numeric,2) as average_sales
from rossmann_sales
group by month
order by month asc""",engine)

print("Average sales per Month")
print(q4)


Average sales per Month
    month  average_sales
0       1        6564.30
1       2        6589.49
2       3        6976.82
3       4        7046.66
4       5        7106.81
5       6        7001.40
6       7        6953.58
7       8        6649.23
8       9        6547.47
9      10        6602.97
10     11        7188.55
11     12        8608.96


It shows the most of the people do shopping in the month of November and December. 

## Store Rankings using RANK()
Advanced window function ranking all 1,115 stores by performance.

In [6]:
q5 = pd.read_sql("""
select 
    store, 
    round(avg(sales)::numeric,2) as average_sales,
    rank() over(order by avg(sales) desc) as sales_rank
from rossmann_sales
group by store
order by average_sales desc
limit 10""",engine)

print("Top 10 store showcasing highest average sales")
print(q5)

Top 10 store showcasing highest average sales
   store  average_sales  sales_rank
0    817       21757.48           1
1    262       20718.52           2
2   1114       20666.56           3
3    251       19123.07           4
4    842       18574.80           5
5    513       18179.09           6
6    562       17969.56           7
7    788       17961.91           8
8    383       17294.72           9
9    756       16574.82          10


Store 817 has highest average sales than any store and it holds rank 1 in the case of sales.

## Month over Month Growth using LAG()
Calculates sales change from previous month using LAG() window function.

In [7]:
q6 = pd.read_sql("""
select 
    month, 
    round(avg(sales)::numeric,1), 
    round(avg(sales) - lag(round(avg(sales),2))
    over(order by month),2) as mom_growth
from rossmann_sales
group by month
order by mom_growth asc""",engine)

print("Sales grown month-over-month")
print(q6)


Sales grown month-over-month
    month   round  mom_growth
0       8  6649.2     -304.35
1       6  7001.4     -105.41
2       9  6547.5     -101.76
3       7  6953.6      -47.82
4       2  6589.5       25.19
5      10  6603.0       55.50
6       5  7106.8       60.15
7       4  7046.7       69.84
8       3  6976.8      387.33
9      11  7188.6      585.58
10     12  8609.0     1420.41
11      1  6564.3         NaN


## Above Average Stores using CTE
Uses two CTEs to identify stores performing above overall average.

In [8]:
q7 =pd.read_sql("""
with average_sales as(
select 
    round(avg(sales)::numeric,2) as overall_average
from rossmann_sales
),

store_sales as (
select store, 
    round(avg(sales)::numeric,2) as store_average
from rossmann_sales
group by store
)

select 
    s.store,
    s.store_average,
    a.overall_average,
    round(s.store_average - a.overall_average::numeric, 2) as difference
from store_sales s
cross join average_sales a
where s.store_average > a.overall_average
order by difference desc
limit 10""",engine)

print("Revenue, average and Revenue of high performing sales store")
print(q7)

Revenue, average and Revenue of high performing sales store
   store  store_average  overall_average  difference
0    817       21757.48          6955.96    14801.52
1    262       20718.52          6955.96    13762.56
2   1114       20666.56          6955.96    13710.60
3    251       19123.07          6955.96    12167.11
4    842       18574.80          6955.96    11618.84
5    513       18179.09          6955.96    11223.13
6    562       17969.56          6955.96    11013.60
7    788       17961.91          6955.96    11005.95
8    383       17294.72          6955.96    10338.76
9    756       16574.82          6955.96     9618.86


## Year-over-Year Growth - Raw Totals (Limited View)
Raw total revenue comparison is included for reference only.
See Query 9 below for the corrected approach using average daily sales,
which accounts for 2015 being a partial year (Jan–Jul only).

In [9]:
q8 =pd.read_sql("""
select 
    year,
    sum(sales) as total_sales, 
    round(sum(sales) - lag(sum(sales))
    over (order by year)::numeric,2) as yoy_growth
from rossmann_sales
group by year
order by year asc""",engine)

print(q8)


   year   total_sales   yoy_growth
0  2013  2.302876e+09          NaN
1  2014  2.180805e+09 -122071188.0
2  2015  1.389500e+09 -791305253.0


In [10]:
q9 =pd.read_sql("""WITH yearly_sales AS (
    SELECT
        year,
        COUNT(DISTINCT date) AS trading_days,
        SUM(sales) AS total_sales,
        ROUND(SUM(sales)::numeric / COUNT(DISTINCT date), 2) AS avg_daily_sales
    FROM rossmann_sales
    GROUP BY year
),
with_growth AS (
    SELECT
        year,
        trading_days,
        total_sales,
        avg_daily_sales,
        ROUND(
            (avg_daily_sales - LAG(avg_daily_sales) OVER (ORDER BY year))
            / LAG(avg_daily_sales) OVER (ORDER BY year) * 100
        , 2) AS daily_avg_growth_pct
    FROM yearly_sales
)
SELECT * FROM with_growth
ORDER BY year""",engine)


print("Sales growth Year on Year")
print(q9)

Sales growth Year on Year
   year  trading_days   total_sales  avg_daily_sales  daily_avg_growth_pct
0  2013           365  2.302876e+09       6309249.55                   NaN
1  2014           365  2.180805e+09       5974807.93                  -5.3
2  2015           212  1.389500e+09       6554243.60                   9.7


Store Performance Tiers (CASE WHEN)

## Store Performance Tier Classification

Segments all 1,115 stores into four performance bands relative 
to the chain-wide average daily sales of €6,955.

| Tier | Threshold | Business Meaning |
|---|---|---|
| Elite | ≥150% of average | Flagship stores, study for replication |
| High Performer | 110–150% | Above average, potential expansion candidates |
| On Track | 90–110% | Within normal operating range |
| Underperformer | <90% of average | Candidates for operational review |

**Finding:** Run `GROUP BY performance_tier` to see distribution 
across the chain.

In [11]:
q10 = pd.read_sql('''
WITH store_metrics AS (
    SELECT
        store,
        ROUND(AVG(sales)::numeric, 2) AS avg_sales,
        COUNT(DISTINCT date) AS active_days,
        SUM(sales) AS total_revenue
    FROM rossmann_sales
    GROUP BY store
),
overall AS (
    SELECT ROUND(AVG(sales)::numeric, 2) AS overall_avg
    FROM rossmann_sales
)
SELECT
    s.store,
    s.avg_sales,
    o.overall_avg,
    ROUND(s.avg_sales - o.overall_avg, 2) AS vs_average,
    ROUND((s.avg_sales - o.overall_avg) / o.overall_avg * 100, 1) AS pct_vs_average,
    CASE
        WHEN s.avg_sales >= o.overall_avg * 1.5 THEN 'Elite'
        WHEN s.avg_sales >= o.overall_avg * 1.1 THEN 'High Performer'
        WHEN s.avg_sales >= o.overall_avg * 0.9 THEN 'On Track'
        ELSE 'Underperformer'
    END AS performance_tier
FROM store_metrics s
CROSS JOIN overall o
ORDER BY s.avg_sales DESC''',engine)

print('Store Performance Tiers')
print(q10)

Store Performance Tiers
      store  avg_sales  overall_avg  vs_average  pct_vs_average  \
0       817   21757.48      6955.96    14801.52           212.8   
1       262   20718.52      6955.96    13762.56           197.9   
2      1114   20666.56      6955.96    13710.60           197.1   
3       251   19123.07      6955.96    12167.11           174.9   
4       842   18574.80      6955.96    11618.84           167.0   
...     ...        ...          ...         ...             ...   
1110    841    2972.61      6955.96    -3983.35           -57.3   
1111    208    2936.29      6955.96    -4019.67           -57.8   
1112    198    2900.60      6955.96    -4055.36           -58.3   
1113    543    2790.38      6955.96    -4165.58           -59.9   
1114    307    2703.74      6955.96    -4252.22           -61.1   

     performance_tier  
0               Elite  
1               Elite  
2               Elite  
3               Elite  
4               Elite  
...               ...  
111

In [12]:
' Tier Distribution Summary '
q11 = pd.read_sql('''
WITH store_metrics AS (
    SELECT store, ROUND(AVG(sales)::numeric, 2) AS avg_sales
    FROM rossmann_sales
    GROUP BY store
),
overall AS (
    SELECT ROUND(AVG(sales)::numeric, 2) AS overall_avg
    FROM rossmann_sales
),
tiered AS (
    SELECT
        CASE
            WHEN s.avg_sales >= o.overall_avg * 1.5 THEN 'Elite'
            WHEN s.avg_sales >= o.overall_avg * 1.1 THEN 'High Performer'
            WHEN s.avg_sales >= o.overall_avg * 0.9 THEN 'On Track'
            ELSE 'Underperformer'
        END AS performance_tier
    FROM store_metrics s CROSS JOIN overall o
)
SELECT performance_tier, COUNT(*) AS store_count,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (), 1) AS pct_of_chain
FROM tiered
GROUP BY performance_tier
ORDER BY store_count DESC''',engine)

print('The Distribution Summary')
print(q11)

The Distribution Summary
  performance_tier  store_count  pct_of_chain
0   Underperformer          487          43.7
1         On Track          300          26.9
2   High Performer          250          22.4
3            Elite           78           7.0


## Promotional Effectiveness by Store Type

**Business Question:** In Q4 (Oct–Dec), when Christmas demand is already 
elevated, do promotions still generate meaningful sales lift? 
And does the answer differ by store type?

This matters for promotional budget allocation — if Store Type B already 
performs at peak in Q4 without promotions, running them may be 
unnecessary spend.

**Finding:** Promotional lift in Q4 is not even across store types, and the pattern is the opposite of what you might expect. Store Type A gets the largest lift from promotions (+38.3%), followed by Type D (+32.9%) and Type C (+27.8%). Store Type B, which already has the highest baseline sales of any format, gets the smallest lift (+19.4%).

This makes sense once you consider what a promo is actually doing: it's pulling in sales that wouldn't have happened otherwise. Type B stores are already performing close to their ceiling in Q4 (the holiday season alone is doing a lot of the work), so there's less room for a promotion to add incremental sales. Type A stores, with a lower baseline, have more headroom for a promo to make a real difference.

**Business implication:** If Q4 promotional budget is currently allocated evenly across store types, or worse, weighted toward the already-strong Type B stores, that's likely an inefficient use of spend. Reallocating a larger share of the promo budget toward Type A and Type D stores in Q4 would be expected to generate more incremental sales per euro spent than maintaining current levels at Type B.

One caveat worth flagging: Type B has only 17 stores in this dataset versus 600+ for Type A, so the Type B lift percentage is based on a much smaller sample and should be treated as a useful signal rather than a precise number.

In [13]:
'Quarter 4 Promotional Effectiveness by Store Type '
q12 = pd.read_sql('''
SELECT
    s.storetype,
    COUNT(DISTINCT r.store) AS store_count,
    ROUND(AVG(CASE WHEN r.promo = 1 THEN r.sales END)::numeric, 2) AS q4_promo_avg,
    ROUND(AVG(CASE WHEN r.promo = 0 THEN r.sales END)::numeric, 2) AS q4_no_promo_avg,
    ROUND(
        (AVG(CASE WHEN r.promo = 1 THEN r.sales END) -
         AVG(CASE WHEN r.promo = 0 THEN r.sales END))
        / AVG(CASE WHEN r.promo = 0 THEN r.sales END) * 100
    , 1) AS q4_promo_lift_pct
FROM rossmann_sales r
JOIN rossmann_store s ON r.store = s.store
WHERE r.month IN (10, 11, 12)
GROUP BY s.storetype
ORDER BY q4_promo_lift_pct DESC''',engine)

print('Quarter 4 Promotional Effectiveness by Store Type')
print(q12)

Quarter 4 Promotional Effectiveness by Store Type
  storetype  store_count  q4_promo_avg  q4_no_promo_avg  q4_promo_lift_pct
0         a          602       8756.52          6333.33               38.3
1         d          348       8442.05          6354.00               32.9
2         c          148       8607.58          6737.19               27.8
3         b           17      11971.18         10022.09               19.4


In [14]:
'''Day-of-week sales, corrected for the number of stores actually trading each day
    This is the fix for the EDA claim that Sunday/Monday are "peak" days.
    Sunday's high average is driven by a small number of stores being open at all,
    not by demand being unusually strong. This query proves it.'''

q13 = pd.read_sql('''
SELECT
    dayofweek,
    COUNT(DISTINCT store) AS stores_open,
    ROUND(AVG(sales)::numeric, 2) AS avg_daily_sales,
    SUM(sales) AS total_sales
FROM rossmann_sales
GROUP BY dayofweek
ORDER BY dayofweek''',engine)

In [15]:
print(q13)

   dayofweek  stores_open  avg_daily_sales   total_sales
0          1         1115          8216.25  1.130203e+09
1          2         1115          7088.41  1.020412e+09
2          3         1115          6728.79  9.549629e+08
3          4         1115          6768.21  9.111777e+08
4          5         1115          7073.03  9.805559e+08
5          6         1115          5875.08  8.463177e+08
6          7           33          8224.72  2.955143e+07
